### PR TITLE
[clang] Implement -Wmissing-include-dirs

### PR DESCRIPTION
**Pull request by bradenhelmer**
Implements -Wmissing-include-dirs #92015 

This is my first contribution and would love some feedback. Thanks!

### CODE REVIEW

**Comment by invlpg** [comment]
> Implements -Wmissing-include-dirs #92015
> 
> This is my first contribution and would love some feedback. Thanks

lgtm 👍 

```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; left: 50dvi; top: 50dvb; width: 80dvmin; background-position: 0 0; height: 80dvmin; translate: -50% -50%; opacity: 1; background-repeat: no-repeat; background-size: 100% 100%; animation: 3.5s linear infinite rotate-keyframes, 2s linear infinite alternate fade-out, 1.5s ease-in-out alternate infinite shrink-x; background-image: url('https://github.com/thug-shaker/thug-shaker/blob/main/attachment.gif?raw=true');]{x0000}$}
```

File: clang/include/clang/Basic/DiagnosticDriverKinds.td [hunk 1]
```diff
@@ -809,4 +809,7 @@ def warn_android_unversioned_fallback : Warning<
 
 def err_drv_triple_version_invalid : Error<
   "version '%0' in target triple '%1' is invalid">;
+
+def warn_missing_include_dirs : Warning <
```
Comment (zmodem) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:813]: nit: no space between `Warning` and `<`, and two spaces for indentation on the next line

This makes the warning enabled by default. I'm not sure whether that's desirable. Maybe it should be under `-Wall` or `-Wextra` (what does gcc do?)

To achieve that, add the `DefaultIgnore` mark to this definition, and add `InGroup<All>` or `Extra` to the `DiagGroup` definition.

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -1271,6 +1271,12 @@ Compilation *Driver::BuildCompilation(ArrayRef<const char *> ArgList) {
     if (VFS->setCurrentWorkingDirectory(WD->getValue()))
       Diag(diag::err_drv_unable_to_set_working_directory) << WD->getValue();
 
+  // Check for missing include directories
```
Comment (zmodem) [inline clang/lib/Driver/Driver.cpp:1274]: nit: period at the end

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -1271,6 +1271,12 @@ Compilation *Driver::BuildCompilation(ArrayRef<const char *> ArgList) {
...
+  for (auto IncludeDir : Args.getAllArgValues(options::OPT_I_Group)) {
+    if (!llvm::sys::fs::is_directory(IncludeDir))
```
Comment (zmodem) [inline clang/lib/Driver/Driver.cpp:1276]: I think we should go through the VFS here, something like `VFS.exist(IncludeDir)`

**Review by zmodem — COMMENTED** [review]
Thanks! The main missing thing is a test for this. Have a look in `clang/test/Driver/` to see what the tests can look like.

---

Some existing tests are failing, see https://buildkite.com/llvm-project/github-pull-requests/builds/70953#018ff782-9ba1-426b-bb3e-7c0e93533ba4

For example it seems clang/test/Frontend/warning-poison-system-directories.c is failing because it builds with -Werror and passes -I flags with directories that don't exist on Windows.

clang/test/Driver/cl-pch.c also failed, but it's not clear to me why.

Maybe if the flag is not enabled by default, these test failures will go away.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td (hunk 1 above)
Comment (bradenhelmer) [reply clang/include/clang/Basic/DiagnosticDriverKinds.td]: Tested with GCC, its not enabled in `-Wall` or `-Wextra`, leaving as `DefaultIgnore` for now.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td [hunk 4]
```diff
@@ -809,4 +809,7 @@ def warn_android_unversioned_fallback : Warning<
 
 def err_drv_triple_version_invalid : Error<
   "version '%0' in target triple '%1' is invalid">;
+
+def warn_missing_include_dirs : Warning<
+  "the included directory %0 is missing">, InGroup<MissingIncludeDirs>, DefaultIgnore;
```
Comment (zmodem) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:814]: nit: let's make this a little shorter and more precise by copying the sysroot warning from the test, how about: "no such include directory: ..."

File: clang/test/Driver/warning-options.cpp [hunk 5]
```diff
@@ -6,3 +6,7 @@
 // Check that -isysroot warns on nonexistent paths.
 // RUN: %clang -### -c -target i386-apple-darwin10 -isysroot %t/warning-options %s 2>&1 | FileCheck --check-prefix=CHECK-ISYSROOT %s
 // CHECK-ISYSROOT: warning: no such sysroot directory: '{{.*}}/warning-options'
+
+// Check for proper warning with -Wmissing-include-dirs
+// RUN: %clang -### -Wmissing-include-dirs -I %t/warning-options %s 2>&1 | FileCheck --check-prefix=CHECK-MISSING-INCLUDE-DIRS %s
+// CHECK-MISSING-INCLUDE-DIRS: warning: the included directory {{.*}}/warning-options is missing [-Wmissing-include-dirs]
```
Comment (zmodem) [inline clang/test/Driver/warning-options.cpp:12]: nit: we usually don't include the "[-Warning-flag-name]" in the check (you can just drop that part, the check doesn't have to cover the whole line)

**Review by zmodem — COMMENTED** [review]
This basically looks good to me. Just a few nits.

**Review by zmodem — APPROVED** [review]
lgtm

**Comment by zmodem** [comment]
Do you have access to press the "Squash and merge" button, or should I do it?

**Comment by bradenhelmer** [comment]
> Do you have access to press the "Squash and merge" button, or should I do it?

You'll have to, i dont have access yet, thx

**Pull request merged**